system/lib/wasmfs/streams.cpp

```diff
@@ -43,20 +44,37 @@ __wasi_errno_t StdoutFile::write(const uint8_t* buf, size_t len, off_t offset) {
   return writeStdBuffer(buf, len, &_emscripten_out, writeBuffer);
 }
 
+void StdoutFile::flush() {
+  // Write a null to flush the output (see comment above on "Flush on either a
+  // null or a newline").
+  const uint8_t nothing = '\0';
+  write(&nothing, 1, 0);
+}
+
 std::shared_ptr<StdoutFile> StdoutFile::getSingleton() {
   static const std::shared_ptr<StdoutFile> stdoutFile =
-    std::make_shared<StdoutFile>(S_IWUGO);
+    std::make_shared<StdoutFile>();
   return stdoutFile;
 }
 
 __wasi_errno_t StderrFile::write(const uint8_t* buf, size_t len, off_t offset) {
   // Similar issue with Node and worker threads as emscripten_out.
+  // TODO: May not want to proxy stderr (fd == 2) to the main thread, as
+  //       emscripten_err does.
+  //       This will not show in HTML - a console.warn in a worker is
+  //       sufficient. This would be a change from the current FS.
   return writeStdBuffer(buf, len, &_emscripten_err, writeBuffer);
 }
 
+void StderrFile::flush() {
+  // Similar to StdoutFile.
+  const uint8_t nothing = '\0';
+  write(&nothing, 1, 0);
```

Comment:
This breaks our locking model a bit by exposing `write` directly instead of through the handle, so there's nothing checking that the lock is held. These methods should either do their own locking, or even better, `flush` should be a protected pure virtual function exposed only via `DataFile::Handle::flush`, just like `read` and `write`.

Reply:
This happens during shutdown so it is safe, but I do see what you mean, yeah...

You're suggesting we want to add a general flush operation on all streams? That seems reasonable, although I hadn't thought that far ahead. Not sure how to proceed here, in that case - we could wait until we have a better idea on persistence and flushing and so forth?

Reply:
Or we could make this a private method for now and add `friend WasmFS`. It's kind of ugly, but I would be fine with that for now.

Reply:
Ok, I added a flush method now and added locking.

I am actually not that happy about the result - it feels weird to see us create and operate on StdoutFile without a lock in `initRootDirectory`, just a few lines lower down than the diff, but then to lock in the destructor even though it is symmetrical in that one place is the start of the entire FS and the other is the end of it all. I added a comment trying to explain the oddity but I'm not sure. What do you think?

Reply:
But the locks are acquired in `initRootDirectory` as well (via `locked()`), so I think this is consistent. I would use `locked` rather than create handles explicitly in the destructor, too.

Reply:
Good points. Changed to use `locked()`.